lib/ClientImpl.cc

```diff
@@ -735,23 +739,14 @@ void ClientImpl::shutdown() {
     auto producers = producers_.move();
     auto consumers = consumers_.move();
 
-    for (auto&& kv : producers) {
-        ProducerImplBasePtr producer = kv.second.lock();
-        if (producer) {
-            producer->shutdown();
-        }
-    }
-
-    for (auto&& kv : consumers) {
-        ConsumerImplBasePtr consumer = kv.second.lock();
-        if (consumer) {
-            consumer->shutdown();
-        }
-    }
+    auto numProducers = producers.size();
+    auto numConsumers = consumers.size();
+    // Call the destructors out of the lock to avoid deadlocks
+    producers.clear();
```

Comment:
The producer in the map `producers` is a weak ptr. Calling the `clear` here doesn't guarantee the desturctor will be called. So the shutdown may not be called here. However, I think we should make sure all the producers/consumers should be shutdown after the client shutdown.

Reply:
Good find! It makes sense. Let me think of a better solution.

Reply:
@RobertIndie I added `shutdown` back as the virtual function and a new non-virtual function `internalShutdown()` is added instead. Could you take a look again?